apiserver/facades/client/bundle/bundle.go

```diff
@@ -210,13 +233,135 @@ func (b *APIv1) GetChanges(args params.BundleChangesParams) (params.BundleChange
 	})
 }
 
-type validators struct {
-	verifyConstraints func(string) error
-	verifyStorage     func(string) error
-	verifyDevices     func(string) error
+// GetChanges returns the list of changes required to deploy the given bundle
+// data. The changes are sorted by requirements, so that they can be applied in
+// order.
+// GetChanges has been superseded in favour of GetChangesMapArgs. It's
+// preferable to use that new method to add new functionality and move clients
+// away from this one.
+func (b *APIv5) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
+	vs := validators{
+		verifyConstraints: func(s string) error {
+			_, err := constraints.Parse(s)
+			return err
+		},
+		verifyStorage: func(s string) error {
+			_, err := storage.ParseConstraints(s)
+			return err
+		},
+		verifyDevices: func(s string) error {
+			_, err := devices.ParseConstraints(s)
+			return err
+		},
+	}
+	return b.getChanges(args, vs, func(changes []bundlechanges.Change, results *params.BundleChangesResults) error {
```

Comment:
Having to duplicate all this code makes it really hard to track what calls what. 

How about we change the `b.getChanges` signature to include a boolean flag (or alternatively, a typed enum for more clarity) which indicates the parse mode (e.g. `includeOverlays`) to use?

Then you can extract the common `GetChanges` logic into a private method which accepts this flag and just provide two tiny stubs for APIV5 and BundleAPI which pass the right value through.

Reply:
I agree, the duplication adds more indirection on the control flow. I initially tried to duplicate only the `b.getChanges` but it didn't work for reasons related to Go semantics that I don't understand yet.

However, if I'm understanding your proposal correctly, I'm also doubtful that adding another mechanic to separate functionality between versions would make this less confusing. Especially when more version breaks are added in the future, this might potentially turn into (already existing) duplicate codes with big switches in some of the functions.

Also, code duplication is often really bad for like maintainance/debug etc, but this duplication is like grabbing a snapshot of the current version and moving on, so day 2 ops are still gonna be done on the `func (b *BundleAPI) GetChanges` (unless it's specifically about `<=APIv5`). On the other hand, adding anohter argument would sort of cram all the logics into one spot, thereby forcing anyone maintaining/improving it in the future to understand all of it and make sure they're not breaking any other version logic.

Reply:
My suggestion was something along these lines:

```
func (b *APIv5) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
 return b.doGetChanges(args, false)
}

// Versions 6+
func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
 return b.doGetChanges(args, true)
}

// You could also pass validators as an arg and use the same method for pre V5 facades
func (b *BundleAPI) doGetChanges(args params.BundleChangesParams, expandOverlays bool) (params.BundleChangesResults, error) {
    vs := validators{
		verifyConstraints: func(s string) error {
			_, err := constraints.Parse(s)
			return err
		},
		verifyStorage: func(s string) error {
			_, err := storage.ParseConstraints(s)
			return err
		},
		verifyDevices: func(s string) error {
			_, err := devices.ParseConstraints(s)
			return err
		},
	}
	return b.getBundleChanges(args, vs, expandOverlays, mapBundleChanges)
}

func mapBundleChanges(changes []bundlechanges.Change, results *params.BundleChangesResults) error {
  results.Changes = make([]*params.BundleChange, len(changes))
  for i, c := range changes {
	  var guiArgs []interface{}
	  switch c := c.(type) {
	  case *bundlechanges.AddApplicationChange:
		  guiArgs = c.GUIArgsWithDevices()
	  default:
		  guiArgs = c.GUIArgs()
	  }
	  results.Changes[i] = &params.BundleChange{
		  Id:       c.Id(),
		  Method:   c.Method(),
		  Args:     guiArgs,
		  Requires: c.Requires(),
	  }
  }
  return nil
}

func (b *BundleAPI) getBundleChanges(...){
  // merge getChanges + getBundleChanges logic
}
```

The way that the facade abuses inheritance should allow this to work but as you can see the code is much more compact. Also, note that when we move to server-side bundle expansion this facade will become obsolete as we will be using the same code path for both the juju cli and pylibjuju and the controller will take care of calculating and applying the required changes.

Reply:
@achilleasa Ah ok I see what you're saying, I agree, this looks much better, I'll see what I can do, thanks for the input :+1:

Reply:
Forgot to add this handy snippet if you want to add some more context instead of passing true/false:

```go
const (
   ignoreOverlays = false
   includeOverlays = true
)

...

func (b *APIv5) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
 return b.doGetChanges(args, ignoreOverlays)
}

// Versions 6+
func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
 return b.doGetChanges(args, includeOverlays)
}
```